### DISCUSSION TITLE
Simulation dans le temps et auto balance ?

### DISCUSSION THREAD

**Discussion by nmathey** [MadeInPierre/finalynx#139] [Q&A] [answered]
Hello,

Ca y est j'ai refait ma structure et tester le simulateur.
Si j'ai bien compris ton exemple, tu mets par défaut toutes tes rentrées d'argents sur ton livret A puis comptes sur l'auto balance afin que le simulateur répartisse "l'épargne" (revenus-dépenses) au fur et à mesure sur les autres enveloppes `Folder` jusqu'à la date de fin?

Si ma compréhension est bonne, je dois avoir un "hic" quelque part car affichant le résultat à cette date de fin j'ai des incohérences: 

- je devrais avoir "plus" : peut-être lié au fait que je suis pessimiste sur l'évolution de revenus (1%) et de l'inflation (3%)?
- certaines "targets" sont "dites OK" alors qu'elles ne le sont pas: mes maximum ne sont pas atteints mais sont vus "OK" (matelas de sécurités et disponibilités dans mon cas)
- l'auto balance ne semble pas faire ce que j'attends (ou je n'ai tout simplement pas compris son fonctionnement :)): cela me "vide" des enveloppes alors que cela ne devrait pas et ne "vide pas" certaines qui le devraient (par exemple dans mon cas, mon matelas de sécurités et disponibilités qui devraient être à leur max, AV - Linxea qui a été vidée et AV - MPP qui a augmenté alors qui devrait être vide)

Mon portefolio à date

```
 ‣ 52185 € My Portfolio                                                         
   ├── ✓ 12000 € 23% Matelas de sécurité - Très court terme - Maximum 12000 €     
   │   └── ‣ 12000 € 100% Livrets                                                 
   ├── ✓  8514 € 16% Disponibilités - Court/Moyen terme - Maximum 24000 €         
   │   ├── ‣ 1017 € 12% Livrets                                                   
   │   ├── ‣ 6739 € 79% PEE                                                       
   │   └── ‣  758 €  9% Fonds Euro - AV                                           
   └── ‣ 31671 € 61% Investissements - Long Terme                                 
       ├── ‣ 23290 € 74% Actions                                                  
       │   ├── ‣ 9084 € 39% CTO - IB                                              
       │   ├── ‣ 1192 €  5% CTO - Revolut                                         
       │   ├── ↘ 2143 €  9% AV - MPP - Maximum 0 €                                
       │   ├── ‣ 1875 €  8% AV - Linxea                                           
       │   └── ‣ 8996 € 39% AV - WeSave                                           
       ├── ‣  6806 € 21% Immobiliers                                              
       │   ├── ‣ 1484 € 22% SCPI - Epsilon360                                     
       │   ├── ‣ 1122 € 16% SCPI - Neo                                            
       │   └── ‣ 4200 € 62% DEFI - RealT                                          
       └── ‣  1575 €  5% Exotiques                                                
           ├── ‣ 750 € 48% Crypto                                                 
           └── ‣ 825 € 52% Startups & PME                                         
               ├── ‣ 500 € 61% Finary                                             
               └── ‣ 325 € 39% Mt Pelerin Shares

```

Mon portfolio simulé en 2050:

```
‣ 112011 € My Portfolio                                                         
   ├── ✓   2459 €  2% Matelas de sécurité - Très court terme - Maximum 12000 €     
   │   └── ‣ 2459 € 100% Livrets                                                   
   ├── ✓   8816 €  8% Disponibilités - Court/Moyen terme - Maximum 24000 €         
   │   ├── ↯    0 €  0% Livrets - Invest!                                          
   │   ├── ‣ 8816 € 100% PEE                                                       
   │   └── ↯    0 €  0% Fonds Euro - AV - Invest!                                  
   └── ‣ 100737 € 90% Investissements - Long Terme                                 
       ├── ‣ 79952 € 79% Actions                                                   
       │   ├── ‣ 33915 € 42% CTO - IB                                              
       │   ├── ‣  4450 €  6% CTO - Revolut                                         
       │   ├── ↘  8001 € 10% AV - MPP - Maximum 0 €                                
       │   ├── ↯     0 €  0% AV - Linxea - Invest!                                 
       │   └── ‣ 33586 € 42% AV - WeSave                                           
       ├── ‣ 20129 € 20% Immobiliers                                               
       │   ├── ‣  2533 € 13% SCPI - Epsilon360                                     
       │   ├── ‣  1915 € 10% SCPI - Neo                                            
       │   └── ‣ 15681 € 78% DEFI - RealT                                          
       └── ‣   656 €  1% Exotiques                                                 
           ├── ‣ 294 € 45% Crypto                                                  
           └── ‣ 362 € 55% Startups & PME                                          
               ├── ‣ 220 € 61% Finary                                              
               └── ‣ 143 € 39% Mt Pelerin Shares   

```

En tout cas bon boulot c'est vraiment top et très complémentaire de Finary et cela remplacera mes propres simulations bancales sous Excel :)

**Comment by nmathey** [comment]
Info supplémentaire: j'ai rajouté des `TargetRatio` sur mes trois actifs d'investissements: à partir de là le simulateur plante et me dis `ValueError: Attempted to remove too much from the bucket.`

En ne lançant pas le simulateur il m'affiche bien le résultat à date.

Donc il doit y avoir un schmilblick avec l'auto balance ^^

**Reply by MadeInPierre** [reply]
Salut ! Bizarrement ma simu fait également des trucs chelous, je ne sais pas d'où ça vient, je ne me souviens pas d'avoir changé quoi que ce soit depuis :sweat_smile: 

<img src="https://github.com/MadeInPierre/finalynx/assets/12670280/0f1ad4e0-6f1f-476b-ba22-39f6d3e6610a" width="600px"/>

Je regarde un peu dès que j'ai le temps et te tiens au courant !

**Reply by MadeInPierre** [reply]
Hey! Je suis en vacances et n'aurai pas trop le temps de regarder quelques semaines, mais j'ai découvert que la version actuelle fonctionne mais est un peu capricieuse. 

En gros il faut que le dossier long terme soit rempli de dossiers et lignes ayant chacun un `TargetRatio`, il ne doit pas y avoir de dossier vide. Pour les dossiers auto-remplis par enveloppe, il ne faut pas qu'il y ait de sous-dossier, et les lignes auto-remplies sont considérées comme ayant un `TargetRatio(1/n)` (e.g. `TargetRatio(25)` s'il y a 4 lignes dans le dossier). Donc en ayant une structure très homogène et propre ça fonctionne :sweat_smile: 

J'espère que ça aide pour l'instant, j'essaierai de faire un truc plus robuste courant septembre :slightly_smiling_face:

**Comment by MadeInPierre** [comment]
> Si j'ai bien compris ton exemple, tu mets par défaut toutes tes rentrées d'argents sur ton livret A puis comptes sur l'auto balance afin que le simulateur répartisse "l'épargne" (revenus-dépenses) au fur et à mesure sur les autres enveloppes Folder jusqu'à la date de fin?

Exact, l'épargne va dans un livret qui est dans un Bucket pour le court/moyen terme. Ensuite, un dernier SharedFolder dans le long terme avec ce qui dépasse des livrets a un `TargetRatio(0)`, comme ça tout ce qui dépasse sera automatiquement investi dans le long terme

> je devrais avoir "plus" : peut-être lié au fait que je suis pessimiste sur l'évolution de revenus (1%) et de l'inflation (3%)?

Si ton graphique a la même forme que ma photo ci-dessus, alors c'est un "bug" car ta structure n'est pas comme ce que j'avais prévu initialement (voir ma réponse ci-dessus). En tout cas, les revenus sont toujours considérés nets d'inflation (avec 2500€ de revenus, le simulateur appliquera encore 2500€ en 2050) et l'évolution est nette d'inflation. Seuls les placements sont impactés par l'inflation (à améliorer ?), e.g. si tu mets 8% de `perf` sur un ETF et inflation de 3%, le placement te rapportera 5%.

> certaines "targets" sont "dites OK" alors qu'elles ne le sont pas: mes maximum ne sont pas atteints mais sont vus "OK" (matelas de sécurités et disponibilités dans mon cas)

Peut-être as-tu inversé `TargetMin` et `TargetMax` ? `TargetMin(12000)` veut dire que tu souhaiterais avoir au moins 12000€ sur ce dossier/ligne, il sera rouge si tu es en dessous.

> l'auto balance ne semble pas faire ce que j'attends (ou je n'ai tout simplement pas compris son fonctionnement :)): cela me "vide" des enveloppes alors que cela ne devrait pas et ne "vide pas" certaines qui le devraient (par exemple dans mon cas, mon matelas de sécurités et disponibilités qui devraient être à leur max, AV - Linxea qui a été vidée et AV - MPP qui a augmenté alors qui devrait être vide)

Je n'arrive pas trop à visualiser comme ça malheureusement ^^' peut-être si j'avais ta config je pourrais me faire une idée, j'espère que mes commentaires pourraient aider à mieux comprendre ce que fair l'algo pour avoir une structure compatible :crossed_fingers:

**Reply by nmathey** [reply]
T'inquiètes pas d'urgence. Je te partagerais ma config à ton retour de repos/vacances ;)
J'ai essayé de rajouter quelques TargetRatio et rajouter le ShareFolder dans le long terme mais rien y fait: maintenant ma config est assez complexe donc on prendra le temps de debugger cela ensemble ^^

**Reply by MadeInPierre** [reply]
Hola! Me revoici ^^ Si jamais la chose t'intéresse toujours, j'ai créé un [tuto pour la structure](https://github.com/MadeInPierre/finalynx/blob/main/examples/tutorials/3_simulation/2_template.py) qui je sais fonctionne pour la simulation. En gros, il faut que le dossier long terme ait uniquement des `TargetRatio` (pas de `TargetMin` ou autres), et qu'il contienne un `SharedFolder` sans aucun `amount` défini avec un `TargetRatio` à 0% avec Livret A dedans. Ce dernier `SharedFolder` va donc contenir tout ce qui dépasse et qui n'a pas été consommé dans d'autres `SharedFolder` plus haut.

Comme ça, le simulateur ajoute ton épargne tous les mois dans ton Livret A, et à chaque arbitrage il verra qu'il faut 0% dans le `SharedFolder`. S'il y a des sous dedans ça veut dire qu'il faut l'investir, et il le distribuera donc dans les autres lignes du long terme.

J'espère que c'est clair et que ça peut aider ?

**Reply by nmathey** [reply]
Hello,

Bien sûr je ne lâche rien. :D
Je pense avoir corrigé quelques déclarations suite à ton tuto qui fait que les surplus des cours et moyens termes arrivent bien sur le long terme.
Maintenant j'ai toujours un problème dès que j'utilise plus qu'un bucket : erreur --> `ValueError: Attempted to remove too much from the bucket.`

Il faudrait limite que je te partage mon fichier car il est devenu structuré (comprendre complexe :p) comme je l'ai en tête ^^

Puis je rajoute une autre question au passage :)
Est-il possible d'avoir plusieurs "entrées" en même temps sur deux buckets différents?
Genre deux "event Salary" sur la même période?
[réponse à moi-même] Oui apparement et cela me permet de rajouter l'apport (que je déduis de mon salaire) + abonnement à mon PEE ^^

**Reply by nmathey** [reply, marked as answer]
Re,

Bon j'ai fais une grosse simplification de ma structure et j'arrive je pense à un résultat viable ^^

<img width="618" alt="Capture d’écran 2023-09-18 à 18 24 50" src="https://github.com/MadeInPierre/finalynx/assets/20896232/3c1deeb2-e31f-427c-8f1f-246636956733">

Je kiffe! Encore bon boulot!!
Bon ca fait rêver en voyant le montant final: maintenant y a plus qu'à se tenir à son plan :p!

**Reply by MadeInPierre** [reply]
Et le montant est net d'inflation :grin: on parle d'euros en 2023, donc en 2046 tu auras beaucoup plus (3%/an en plus) :wink: 

Cool !! Je kiffe ton résultat :D Bon bien sûr il manque tous les événements de vie (acheter un maison) et dépenses exceptionnelles (mariage, voiture, ...), prochaine feature quand possible... ça permettrait de connaître le "coût d'opportunité" d'un événement (combien de pertes en bourse si j'achète une voiture dans 2 ans ?)

Tu as repéré des choses à améliorer quelque part ? A la fois code et tutoriels/docs ? J'espère que tu pourras avoir une config qui correspond à ce que tu voulais au départ. Pour inspiration j'ai finalement publié [ma config perso](https://github.com/MadeInPierre/finalynx/blob/main/examples/full_example.py) pour inspiration ^^

> Est-il possible d'avoir plusieurs "entrées" en même temps sur deux buckets différents?

Chaque événement (salaire, arbitrage, performance annuelle, etc) sont totalement indépendants, il peut y avoir plusieurs événements avec la même date d'exécution. Je ne sais juste pas lequel s'exécutera en premier dans ce cas. Ici : 
- Les événements Salaire ajoutent de l'argent à une ligne, 
- A chaque trimestre l'événement arbitrage redistribue tout selon les objectifs,
- A la fin de l'année on ajoute la performance annuelle moins l'inflation,
- Et rebelotte

**Reply by MadeInPierre** [reply]
(P.S. Le thème couleurs te convient ? Tu as essayé `python config.py -t dark` ? S'il y a un soucis n'hésite pas à me dire)

**Comment by nmathey** [comment]
Hello,

> Tu as repéré des choses à améliorer quelque part ? A la fois code et tutoriels/docs ?

Peut être plus d'exemples dans la doc pour chaque "objet": je me note ca en todo pour te proposer mes rajouts ;)

>J 'espère que tu pourras avoir une config qui correspond à ce que tu voulais au départ. 

Oui oui tout à fait. En fait je me suis compliqué la vie en pensant me la simplifier en faisant plein de buckets et enveloppes mais au final ca n'aidait pas vraiment.^^

> Pour inspiration j'ai finalement publié [ma config perso] https://github.com/MadeInPierre/finalynx/blob/main/examples/full_example.py) pour inspiration

Top je vais surtout m'en inspirer pour faire un fichier avec des incrémentations correctes :p 

Je vais maintenant finir de rajouter quelques events pour finaliser ma simulation. Et j'aurai fait le tour jusqu'à la prochaine fois :D

>Le thème couleurs te convient ?

Yes je me suis "amusé" à faire mon custom: j'aime bien le theme sombre :)
